In [0]:
# ========================================
# EXPLORATORY NOTEBOOK
# DATASET: gold_fact_sales
# ========================================


# ========================================
# 0. CONFIGURATION
# ========================================

CATALOG = "ptfrozenfoods_dev"
SOURCE_SCHEMA = "silver"
TARGET_SCHEMA = "gold"

DOMAIN = "facts"
DATASET = "gold_fact_sales"

STORAGE_ACCOUNT = "stptfrozenfoodsdevwe01"
SILVER_CONTAINER = "silver"
GOLD_CONTAINER = "gold"

ORDER_ITEMS_DATASET = "erp_order_items"
ORDERS_CUSTOMERS_DATASET = "silver_orders_customers"
PRODUCTS_DATASET = "erp_products"
SALES_CHANNELS_DATASET = "reference_sales_channels"
CALENDAR_DATASET = "reference_calendar"

ORDER_ITEMS_TABLE = f"{CATALOG}.{SOURCE_SCHEMA}.{ORDER_ITEMS_DATASET}"
ORDERS_CUSTOMERS_TABLE = f"{CATALOG}.{SOURCE_SCHEMA}.{ORDERS_CUSTOMERS_DATASET}"
PRODUCTS_TABLE = f"{CATALOG}.{SOURCE_SCHEMA}.{PRODUCTS_DATASET}"
SALES_CHANNELS_TABLE = f"{CATALOG}.{SOURCE_SCHEMA}.{SALES_CHANNELS_DATASET}"
CALENDAR_TABLE = f"{CATALOG}.{SOURCE_SCHEMA}.{CALENDAR_DATASET}"

CANDIDATE_TARGET_TABLE = f"{CATALOG}.{TARGET_SCHEMA}.fact_sales"

SILVER_BASE_PATH = f"abfss://{SILVER_CONTAINER}@{STORAGE_ACCOUNT}.dfs.core.windows.net/"
GOLD_BASE_PATH = f"abfss://{GOLD_CONTAINER}@{STORAGE_ACCOUNT}.dfs.core.windows.net/"

ORDER_ITEMS_SOURCE_PATH = f"{SILVER_BASE_PATH}erp/{ORDER_ITEMS_DATASET}/"
ORDERS_CUSTOMERS_SOURCE_PATH = f"{SILVER_BASE_PATH}integration/silver_orders_customers/"
PRODUCTS_SOURCE_PATH = f"{SILVER_BASE_PATH}erp/{PRODUCTS_DATASET}/"
SALES_CHANNELS_SOURCE_PATH = f"{SILVER_BASE_PATH}reference/{SALES_CHANNELS_DATASET}/"
CALENDAR_SOURCE_PATH = f"{SILVER_BASE_PATH}reference/{CALENDAR_DATASET}/"

In [0]:
# ========================================
# 1. CONTEXT SETUP
# ========================================

print("=" * 80)
print("STARTING GOLD EXPLORATORY: gold_fact_sales")
print("=" * 80)

# Set catalog
spark.sql(f"USE CATALOG {CATALOG}")

# Set source schema
spark.sql(f"USE SCHEMA {SOURCE_SCHEMA}")

print("[INFO] Context setup completed successfully.")

STARTING GOLD EXPLORATORY: gold_fact_sales
[INFO] Context setup completed successfully.


In [0]:
# ========================================
# 2. CONFIGURATION SUMMARY
# ========================================

print("=" * 80)
print("GOLD EXPLORATORY NOTEBOOK CONFIGURATION")
print("=" * 80)
print(f"Catalog:                         {CATALOG}")
print(f"Source schema:                   {SOURCE_SCHEMA}")
print(f"Target schema:                   {TARGET_SCHEMA}")
print(f"Domain:                          {DOMAIN}")
print(f"Dataset:                         {DATASET}")
print(f"Order items table:               {ORDER_ITEMS_TABLE}")
print(f"Orders customers table:          {ORDERS_CUSTOMERS_TABLE}")
print(f"Products table:                  {PRODUCTS_TABLE}")
print(f"Sales channels table:            {SALES_CHANNELS_TABLE}")
print(f"Calendar table:                  {CALENDAR_TABLE}")
print(f"Candidate target table:          {CANDIDATE_TARGET_TABLE}")
print(f"Order items source path:         {ORDER_ITEMS_SOURCE_PATH}")
print(f"Orders customers source path:    {ORDERS_CUSTOMERS_SOURCE_PATH}")
print(f"Products source path:            {PRODUCTS_SOURCE_PATH}")
print(f"Sales channels source path:      {SALES_CHANNELS_SOURCE_PATH}")
print(f"Calendar source path:            {CALENDAR_SOURCE_PATH}")
print("=" * 80)

GOLD EXPLORATORY NOTEBOOK CONFIGURATION
Catalog:                         ptfrozenfoods_dev
Source schema:                   silver
Target schema:                   gold
Domain:                          facts
Dataset:                         gold_fact_sales
Order items table:               ptfrozenfoods_dev.silver.erp_order_items
Orders customers table:          ptfrozenfoods_dev.silver.silver_orders_customers
Products table:                  ptfrozenfoods_dev.silver.erp_products
Sales channels table:            ptfrozenfoods_dev.silver.reference_sales_channels
Calendar table:                  ptfrozenfoods_dev.silver.reference_calendar
Candidate target table:          ptfrozenfoods_dev.gold.fact_sales
Order items source path:         abfss://silver@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/
Orders customers source path:    abfss://silver@stptfrozenfoodsdevwe01.dfs.core.windows.net/integration/silver_orders_customers/
Products source path:            abfss://silver

In [0]:
# ========================================
# 3. PRE-CHECKS
# ========================================

print("[INFO] Checking source table availability...")
spark.sql(f"DESCRIBE TABLE {ORDER_ITEMS_TABLE}")
spark.sql(f"DESCRIBE TABLE {ORDERS_CUSTOMERS_TABLE}")
spark.sql(f"DESCRIBE TABLE {PRODUCTS_TABLE}")
spark.sql(f"DESCRIBE TABLE {SALES_CHANNELS_TABLE}")
spark.sql(f"DESCRIBE TABLE {CALENDAR_TABLE}")

print("[INFO] Checking source path access...")
dbutils.fs.ls(ORDER_ITEMS_SOURCE_PATH)
dbutils.fs.ls(ORDERS_CUSTOMERS_SOURCE_PATH)
dbutils.fs.ls(PRODUCTS_SOURCE_PATH)
dbutils.fs.ls(SALES_CHANNELS_SOURCE_PATH)
dbutils.fs.ls(CALENDAR_SOURCE_PATH)

print("[INFO] Checking target container access...")
dbutils.fs.ls(f"abfss://{GOLD_CONTAINER}@{STORAGE_ACCOUNT}.dfs.core.windows.net/")

print("[INFO] Pre-checks completed successfully.")

[INFO] Checking source table availability...
[INFO] Checking source path access...
[INFO] Checking target container access...
[INFO] Pre-checks completed successfully.


In [0]:
# ========================================
# 4. READ SOURCE DATA
# ========================================

print("[INFO] Loading source tables from the Silver layer...")

# Read source datasets
df_order_items = spark.table(ORDER_ITEMS_TABLE)
df_orders_customers = spark.table(ORDERS_CUSTOMERS_TABLE)
df_products = spark.table(PRODUCTS_TABLE)
df_sales_channels = spark.table(SALES_CHANNELS_TABLE)
df_calendar = spark.table(CALENDAR_TABLE)

print("[INFO] Source tables loaded successfully.")

# Display row counts
print("=" * 80)
print("SOURCE DATA ROW COUNTS")
print("=" * 80)
print(f"Order Items:        {df_order_items.count():,}")
print(f"Orders Customers:   {df_orders_customers.count():,}")
print(f"Products:           {df_products.count():,}")
print(f"Sales Channels:     {df_sales_channels.count():,}")
print(f"Calendar:           {df_calendar.count():,}")
print("=" * 80)

# ========================================
# DATASET PREVIEW — INITIAL EXPLORATION
# ========================================

print("=" * 80)
print("DATASET PREVIEW — INITIAL EXPLORATION")
print("=" * 80)
print("[INFO] Displaying sample data from all source datasets...")
print("[INFO] Each preview shows the first 5 records to validate schema and structure.")
print("=" * 80)


# --------------------------------------------------
# ORDER ITEMS
# --------------------------------------------------
print("\n[INFO] Preview: ORDER ITEMS (df_order_items)")
print("-" * 80)
display(df_order_items.limit(5))


# --------------------------------------------------
# ORDERS AND CUSTOMERS
# --------------------------------------------------
print("\n[INFO] Preview: ORDERS AND CUSTOMERS (df_orders_customers)")
print("-" * 80)
display(df_orders_customers.limit(5))


# --------------------------------------------------
# PRODUCTS
# --------------------------------------------------
print("\n[INFO] Preview: PRODUCTS (df_products)")
print("-" * 80)
display(df_products.limit(5))


# --------------------------------------------------
# SALES CHANNELS
# --------------------------------------------------
print("\n[INFO] Preview: SALES CHANNELS (df_sales_channels)")
print("-" * 80)
display(df_sales_channels.limit(5))


# --------------------------------------------------
# CALENDAR
# --------------------------------------------------
print("\n[INFO] Preview: CALENDAR (df_calendar)")
print("-" * 80)
display(df_calendar.limit(5))


print("\n" + "=" * 80)
print("[INFO] Dataset preview completed successfully.")
print("=" * 80)

[INFO] Loading source tables from the Silver layer...
[INFO] Source tables loaded successfully.
SOURCE DATA ROW COUNTS
Order Items:        245,287
Orders Customers:   89,600
Products:           36
Sales Channels:     4
Calendar:           1,885
DATASET PREVIEW — INITIAL EXPLORATION
[INFO] Displaying sample data from all source datasets...
[INFO] Each preview shows the first 5 records to validate schema and structure.

[INFO] Preview: ORDER ITEMS (df_order_items)
--------------------------------------------------------------------------------


item_pedido_id,pedido_id,produto_id,quantidade,preco_lista_unitario,desconto_unitario,preco_venda_unitario,custo_unitario,lote_fornecedor,flag_promocao,nota_item,load_date,ingestion_timestamp,source_file
IT000000027,PED2021011100012,P012,2,5.49,0.0,5.49,3.14,L66962,0,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv
IT000000061,PED2021011100028,P004,2,8.02,0.0,8.02,5.23,L21140,0,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv
IT000000087,PED2021011200042,P012,2,5.96,0.48,5.48,3.22,L20459,1,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv
IT000000946,PED2021012200421,P009,1,2.37,0.05,2.32,1.12,L26896,0,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv
IT000001177,PED2021012600517,P004,2,7.99,0.0,7.99,4.82,L83053,0,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv



[INFO] Preview: ORDERS AND CUSTOMERS (df_orders_customers)
--------------------------------------------------------------------------------


cliente_id,pedido_id,data_pedido,canal_id,vendedor_id,cidade_entrega,estado_pedido,prazo_entrega_dias,observacao_pedido,sistema_origem,usuario_ultima_alteracao,order_load_date,order_ingestion_timestamp,order_source_file,nome_cliente,tipo_cliente,nif,data_registo,cliente_cidade,distrito,canal_captacao,score_atividade,obs_comercial,segmento,potencial_valor,cluster_comercial,status_cliente,data_status,motivo_status
C0136,PED2021011100001,2021-01-11,CH03,VEND008,Espinho,Entregue,2,null,ERP_FROZEN_V2,maria.s,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Restaurante 136,Restaurante,243297765,2021-01-11,Espinho,Aveiro,Telefone,0.996,null,Pequeno,Alto,Cluster C,Ativo,2026-02-28,null
C0136,PED2021011100002,2021-01-11,CH02,VEND001,Espinho,Entregue,2,null,ERP_FROZEN_V2,maria.s,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Restaurante 136,Restaurante,243297765,2021-01-11,Espinho,Aveiro,Telefone,0.996,null,Pequeno,Alto,Cluster C,Ativo,2026-02-28,null
C0136,PED2021011100003,2021-01-11,CH03,null,Espinho,Entregue,2,null,ERP_FROZEN_V2,ana.c,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Restaurante 136,Restaurante,243297765,2021-01-11,Espinho,Aveiro,Telefone,0.996,null,Pequeno,Alto,Cluster C,Ativo,2026-02-28,null
C0136,PED2021011100004,2021-01-11,CH01,VEND008,Espinho,Entregue,2,null,ERP_FROZEN_V2,svc_sync,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Restaurante 136,Restaurante,243297765,2021-01-11,Espinho,Aveiro,Telefone,0.996,null,Pequeno,Alto,Cluster C,Ativo,2026-02-28,null
C0136,PED2021011100007,2021-01-11,CH03,VEND001,Espinho,Entregue,3,null,ERP_FROZEN_V2,svc_sync,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Restaurante 136,Restaurante,243297765,2021-01-11,Espinho,Aveiro,Telefone,0.996,null,Pequeno,Alto,Cluster C,Ativo,2026-02-28,null



[INFO] Preview: PRODUCTS (df_products)
--------------------------------------------------------------------------------


produto_id,produto_nome,categoria,marca,peso_gramas,preco_lista_base,custo_base_unitario,fornecedor_id,data_lancamento,data_fim,status_produto,popularidade_base,sensibilidade_promocao,fator_sazonal_proprio,codigo_barra_legacy,observacao_interna,load_date,ingestion_timestamp,source_file
P036,Atum em óleo 120g,Conservas & Secos,Mar Azul,120,1.1,0.58,F003,2023-09-08,null,Ativo,0.431,0.893,0.902,56000564949,null,2026-03-17,2026-04-03T22:27:58.849Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_products/load_date=2026-03-17/erp_products.csv
P015,Lasanha bolonhesa 1kg,Pré-cozinhados,Atlântico Select,1000,6.9,4.25,F001,2022-03-03,null,Ativo,2.302,1.291,1.398,56000308650,sku antigo,2026-03-17,2026-04-03T22:27:58.849Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_products/load_date=2026-03-17/erp_products.csv
P022,Pão de cereais 2kg,Padaria,Doce Norte,2000,4.4,2.1,F005,2022-08-23,null,Ativo,0.715,0.724,0.836,56000153631,sku antigo,2026-03-17,2026-04-03T22:27:58.849Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_products/load_date=2026-03-17/erp_products.csv
P026,Gelado baunilha 2L 2kg,Sobremesas,PT Frozen Foods,2000,5.9,2.95,F001,2022-04-25,null,Ativo,0.599,1.2,0.7,56000151954,sku antigo,2026-03-17,2026-04-03T22:27:58.849Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_products/load_date=2026-03-17/erp_products.csv
P021,Pão de água 2400g,Padaria,Padaria Premium,2400,3.9,1.78,F005,2022-10-15,null,Ativo,0.884,1.744,0.789,56000284422,null,2026-03-17,2026-04-03T22:27:58.849Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_products/load_date=2026-03-17/erp_products.csv



[INFO] Preview: SALES CHANNELS (df_sales_channels)
--------------------------------------------------------------------------------


canal_id,nome_canal,descricao,load_date,ingestion_timestamp,source_file
CH02,Vendas Externas,Equipa comercial visitando clientes,2026-03-19,2026-04-02T16:53:14.226Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_sales_channels/load_date=2026-03-19/reference_sales_channels_20260319T151315Z.csv
CH04,Marketplace,Plataformas externas de venda,2026-03-19,2026-04-02T16:53:14.226Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_sales_channels/load_date=2026-03-19/reference_sales_channels_20260319T151315Z.csv
CH03,Telefone,Pedidos feitos por telefone,2026-03-19,2026-04-02T16:53:14.226Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_sales_channels/load_date=2026-03-19/reference_sales_channels_20260319T151315Z.csv
CH01,E-commerce,Pedidos realizados pelo site B2B/B2C,2026-03-19,2026-04-02T16:53:14.226Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_sales_channels/load_date=2026-03-19/reference_sales_channels_20260319T151315Z.csv



[INFO] Preview: CALENDAR (df_calendar)
--------------------------------------------------------------------------------


data,ano,mes,dia,trimestre,nome_mes,dia_semana,nome_dia_semana,is_fim_de_semana,is_inicio_mes,is_fim_mes,load_date,ingestion_timestamp,source_file
2021-01-13,2021,1,13,1,Janeiro,3,Quarta-feira,0,0,0,2026-03-19,2026-04-01T22:04:54.555Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_calendar/load_date=2026-03-19/reference_calendar_20260319T151315Z.csv
2021-01-16,2021,1,16,1,Janeiro,6,Sábado,1,0,0,2026-03-19,2026-04-01T22:04:54.555Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_calendar/load_date=2026-03-19/reference_calendar_20260319T151315Z.csv
2021-01-27,2021,1,27,1,Janeiro,3,Quarta-feira,0,0,1,2026-03-19,2026-04-01T22:04:54.555Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_calendar/load_date=2026-03-19/reference_calendar_20260319T151315Z.csv
2021-02-04,2021,2,4,1,Fevereiro,4,Quinta-feira,0,1,0,2026-03-19,2026-04-01T22:04:54.555Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_calendar/load_date=2026-03-19/reference_calendar_20260319T151315Z.csv
2021-04-05,2021,4,5,2,Abril,1,Segunda-feira,0,1,0,2026-03-19,2026-04-01T22:04:54.555Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_calendar/load_date=2026-03-19/reference_calendar_20260319T151315Z.csv



[INFO] Dataset preview completed successfully.


In [0]:
# ========================================
# 5. INITIAL DATA OVERVIEW
# ========================================

from pyspark.sql import functions as F

print("=" * 80)
print("INITIAL DATA OVERVIEW")
print("=" * 80)

datasets = {
    "Order Items": df_order_items,
    "Orders Customers": df_orders_customers,
    "Products": df_products,
    "Sales Channels": df_sales_channels,
    "Calendar": df_calendar
}

for name, df in datasets.items():
    print("\n" + "=" * 80)
    print(f"DATASET: {name}")
    print("=" * 80)
    
    # Row count
    row_count = df.count()
    print(f"Row Count: {row_count:,}")
    
    # Column count
    column_count = len(df.columns)
    print(f"Column Count: {column_count}")
    
    # Schema
    print("\nSchema:")
    df.printSchema()
    
    # Preview
    print("\nSample Data:")
    display(df.limit(5))
    
    # Null analysis
    print("\nNull Value Analysis:")
    null_counts = df.select([
        F.count(F.when(F.col(c).isNull(), c)).alias(c)
        for c in df.columns
    ])
    display(null_counts)

print("\n[INFO] Initial data overview completed successfully.")

INITIAL DATA OVERVIEW

DATASET: Order Items
Row Count: 245,287
Column Count: 14

Schema:
root
 |-- item_pedido_id: string (nullable = true)
 |-- pedido_id: string (nullable = true)
 |-- produto_id: string (nullable = true)
 |-- quantidade: integer (nullable = true)
 |-- preco_lista_unitario: double (nullable = true)
 |-- desconto_unitario: double (nullable = true)
 |-- preco_venda_unitario: double (nullable = true)
 |-- custo_unitario: double (nullable = true)
 |-- lote_fornecedor: string (nullable = true)
 |-- flag_promocao: integer (nullable = true)
 |-- nota_item: string (nullable = true)
 |-- load_date: date (nullable = true)
 |-- ingestion_timestamp: timestamp (nullable = true)
 |-- source_file: string (nullable = true)


Sample Data:


item_pedido_id,pedido_id,produto_id,quantidade,preco_lista_unitario,desconto_unitario,preco_venda_unitario,custo_unitario,lote_fornecedor,flag_promocao,nota_item,load_date,ingestion_timestamp,source_file
IT000000027,PED2021011100012,P012,2,5.49,0.0,5.49,3.14,L66962,0,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv
IT000000061,PED2021011100028,P004,2,8.02,0.0,8.02,5.23,L21140,0,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv
IT000000087,PED2021011200042,P012,2,5.96,0.48,5.48,3.22,L20459,1,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv
IT000000946,PED2021012200421,P009,1,2.37,0.05,2.32,1.12,L26896,0,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv
IT000001177,PED2021012600517,P004,2,7.99,0.0,7.99,4.82,L83053,0,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv



Null Value Analysis:


item_pedido_id,pedido_id,produto_id,quantidade,preco_lista_unitario,desconto_unitario,preco_venda_unitario,custo_unitario,lote_fornecedor,flag_promocao,nota_item,load_date,ingestion_timestamp,source_file
0,0,930,0,0,0,0,0,0,0,216011,0,0,0



DATASET: Orders Customers
Row Count: 89,600
Column Count: 29

Schema:
root
 |-- cliente_id: string (nullable = true)
 |-- pedido_id: string (nullable = true)
 |-- data_pedido: date (nullable = true)
 |-- canal_id: string (nullable = true)
 |-- vendedor_id: string (nullable = true)
 |-- cidade_entrega: string (nullable = true)
 |-- estado_pedido: string (nullable = true)
 |-- prazo_entrega_dias: integer (nullable = true)
 |-- observacao_pedido: string (nullable = true)
 |-- sistema_origem: string (nullable = true)
 |-- usuario_ultima_alteracao: string (nullable = true)
 |-- order_load_date: date (nullable = true)
 |-- order_ingestion_timestamp: timestamp (nullable = true)
 |-- order_source_file: string (nullable = true)
 |-- nome_cliente: string (nullable = true)
 |-- tipo_cliente: string (nullable = true)
 |-- nif: string (nullable = true)
 |-- data_registo: date (nullable = true)
 |-- cliente_cidade: string (nullable = true)
 |-- distrito: string (nullable = true)
 |-- canal_captacao

cliente_id,pedido_id,data_pedido,canal_id,vendedor_id,cidade_entrega,estado_pedido,prazo_entrega_dias,observacao_pedido,sistema_origem,usuario_ultima_alteracao,order_load_date,order_ingestion_timestamp,order_source_file,nome_cliente,tipo_cliente,nif,data_registo,cliente_cidade,distrito,canal_captacao,score_atividade,obs_comercial,segmento,potencial_valor,cluster_comercial,status_cliente,data_status,motivo_status
C0136,PED2021011100001,2021-01-11,CH03,VEND008,Espinho,Entregue,2,null,ERP_FROZEN_V2,maria.s,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Restaurante 136,Restaurante,243297765,2021-01-11,Espinho,Aveiro,Telefone,0.996,null,Pequeno,Alto,Cluster C,Ativo,2026-02-28,null
C0136,PED2021011100002,2021-01-11,CH02,VEND001,Espinho,Entregue,2,null,ERP_FROZEN_V2,maria.s,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Restaurante 136,Restaurante,243297765,2021-01-11,Espinho,Aveiro,Telefone,0.996,null,Pequeno,Alto,Cluster C,Ativo,2026-02-28,null
C0136,PED2021011100003,2021-01-11,CH03,null,Espinho,Entregue,2,null,ERP_FROZEN_V2,ana.c,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Restaurante 136,Restaurante,243297765,2021-01-11,Espinho,Aveiro,Telefone,0.996,null,Pequeno,Alto,Cluster C,Ativo,2026-02-28,null
C0136,PED2021011100004,2021-01-11,CH01,VEND008,Espinho,Entregue,2,null,ERP_FROZEN_V2,svc_sync,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Restaurante 136,Restaurante,243297765,2021-01-11,Espinho,Aveiro,Telefone,0.996,null,Pequeno,Alto,Cluster C,Ativo,2026-02-28,null
C0136,PED2021011100007,2021-01-11,CH03,VEND001,Espinho,Entregue,3,null,ERP_FROZEN_V2,svc_sync,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Restaurante 136,Restaurante,243297765,2021-01-11,Espinho,Aveiro,Telefone,0.996,null,Pequeno,Alto,Cluster C,Ativo,2026-02-28,null



Null Value Analysis:


cliente_id,pedido_id,data_pedido,canal_id,vendedor_id,cidade_entrega,estado_pedido,prazo_entrega_dias,observacao_pedido,sistema_origem,usuario_ultima_alteracao,order_load_date,order_ingestion_timestamp,order_source_file,nome_cliente,tipo_cliente,nif,data_registo,cliente_cidade,distrito,canal_captacao,score_atividade,obs_comercial,segmento,potencial_valor,cluster_comercial,status_cliente,data_status,motivo_status
0,0,0,0,33760,0,0,0,71679,0,0,0,0,0,0,0,0,0,0,0,0,0,71130,0,0,0,0,0,61150



DATASET: Products
Row Count: 36
Column Count: 19

Schema:
root
 |-- produto_id: string (nullable = true)
 |-- produto_nome: string (nullable = true)
 |-- categoria: string (nullable = true)
 |-- marca: string (nullable = true)
 |-- peso_gramas: integer (nullable = true)
 |-- preco_lista_base: double (nullable = true)
 |-- custo_base_unitario: double (nullable = true)
 |-- fornecedor_id: string (nullable = true)
 |-- data_lancamento: date (nullable = true)
 |-- data_fim: date (nullable = true)
 |-- status_produto: string (nullable = true)
 |-- popularidade_base: double (nullable = true)
 |-- sensibilidade_promocao: double (nullable = true)
 |-- fator_sazonal_proprio: double (nullable = true)
 |-- codigo_barra_legacy: string (nullable = true)
 |-- observacao_interna: string (nullable = true)
 |-- load_date: date (nullable = true)
 |-- ingestion_timestamp: timestamp (nullable = true)
 |-- source_file: string (nullable = true)


Sample Data:


produto_id,produto_nome,categoria,marca,peso_gramas,preco_lista_base,custo_base_unitario,fornecedor_id,data_lancamento,data_fim,status_produto,popularidade_base,sensibilidade_promocao,fator_sazonal_proprio,codigo_barra_legacy,observacao_interna,load_date,ingestion_timestamp,source_file
P036,Atum em óleo 120g,Conservas & Secos,Mar Azul,120,1.1,0.58,F003,2023-09-08,null,Ativo,0.431,0.893,0.902,56000564949,null,2026-03-17,2026-04-03T22:27:58.849Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_products/load_date=2026-03-17/erp_products.csv
P015,Lasanha bolonhesa 1kg,Pré-cozinhados,Atlântico Select,1000,6.9,4.25,F001,2022-03-03,null,Ativo,2.302,1.291,1.398,56000308650,sku antigo,2026-03-17,2026-04-03T22:27:58.849Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_products/load_date=2026-03-17/erp_products.csv
P022,Pão de cereais 2kg,Padaria,Doce Norte,2000,4.4,2.1,F005,2022-08-23,null,Ativo,0.715,0.724,0.836,56000153631,sku antigo,2026-03-17,2026-04-03T22:27:58.849Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_products/load_date=2026-03-17/erp_products.csv
P026,Gelado baunilha 2L 2kg,Sobremesas,PT Frozen Foods,2000,5.9,2.95,F001,2022-04-25,null,Ativo,0.599,1.2,0.7,56000151954,sku antigo,2026-03-17,2026-04-03T22:27:58.849Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_products/load_date=2026-03-17/erp_products.csv
P021,Pão de água 2400g,Padaria,Padaria Premium,2400,3.9,1.78,F005,2022-10-15,null,Ativo,0.884,1.744,0.789,56000284422,null,2026-03-17,2026-04-03T22:27:58.849Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_products/load_date=2026-03-17/erp_products.csv



Null Value Analysis:


produto_id,produto_nome,categoria,marca,peso_gramas,preco_lista_base,custo_base_unitario,fornecedor_id,data_lancamento,data_fim,status_produto,popularidade_base,sensibilidade_promocao,fator_sazonal_proprio,codigo_barra_legacy,observacao_interna,load_date,ingestion_timestamp,source_file
0,0,0,0,0,0,0,0,0,31,0,0,0,0,0,20,0,0,0



DATASET: Sales Channels
Row Count: 4
Column Count: 6

Schema:
root
 |-- canal_id: string (nullable = true)
 |-- nome_canal: string (nullable = true)
 |-- descricao: string (nullable = true)
 |-- load_date: date (nullable = true)
 |-- ingestion_timestamp: timestamp (nullable = true)
 |-- source_file: string (nullable = true)


Sample Data:


canal_id,nome_canal,descricao,load_date,ingestion_timestamp,source_file
CH02,Vendas Externas,Equipa comercial visitando clientes,2026-03-19,2026-04-02T16:53:14.226Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_sales_channels/load_date=2026-03-19/reference_sales_channels_20260319T151315Z.csv
CH04,Marketplace,Plataformas externas de venda,2026-03-19,2026-04-02T16:53:14.226Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_sales_channels/load_date=2026-03-19/reference_sales_channels_20260319T151315Z.csv
CH03,Telefone,Pedidos feitos por telefone,2026-03-19,2026-04-02T16:53:14.226Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_sales_channels/load_date=2026-03-19/reference_sales_channels_20260319T151315Z.csv
CH01,E-commerce,Pedidos realizados pelo site B2B/B2C,2026-03-19,2026-04-02T16:53:14.226Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_sales_channels/load_date=2026-03-19/reference_sales_channels_20260319T151315Z.csv



Null Value Analysis:


canal_id,nome_canal,descricao,load_date,ingestion_timestamp,source_file
0,0,0,0,0,0



DATASET: Calendar
Row Count: 1,885
Column Count: 14

Schema:
root
 |-- data: date (nullable = true)
 |-- ano: integer (nullable = true)
 |-- mes: integer (nullable = true)
 |-- dia: integer (nullable = true)
 |-- trimestre: integer (nullable = true)
 |-- nome_mes: string (nullable = true)
 |-- dia_semana: integer (nullable = true)
 |-- nome_dia_semana: string (nullable = true)
 |-- is_fim_de_semana: integer (nullable = true)
 |-- is_inicio_mes: integer (nullable = true)
 |-- is_fim_mes: integer (nullable = true)
 |-- load_date: date (nullable = true)
 |-- ingestion_timestamp: timestamp (nullable = true)
 |-- source_file: string (nullable = true)


Sample Data:


data,ano,mes,dia,trimestre,nome_mes,dia_semana,nome_dia_semana,is_fim_de_semana,is_inicio_mes,is_fim_mes,load_date,ingestion_timestamp,source_file
2021-01-13,2021,1,13,1,Janeiro,3,Quarta-feira,0,0,0,2026-03-19,2026-04-01T22:04:54.555Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_calendar/load_date=2026-03-19/reference_calendar_20260319T151315Z.csv
2021-01-16,2021,1,16,1,Janeiro,6,Sábado,1,0,0,2026-03-19,2026-04-01T22:04:54.555Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_calendar/load_date=2026-03-19/reference_calendar_20260319T151315Z.csv
2021-01-27,2021,1,27,1,Janeiro,3,Quarta-feira,0,0,1,2026-03-19,2026-04-01T22:04:54.555Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_calendar/load_date=2026-03-19/reference_calendar_20260319T151315Z.csv
2021-02-04,2021,2,4,1,Fevereiro,4,Quinta-feira,0,1,0,2026-03-19,2026-04-01T22:04:54.555Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_calendar/load_date=2026-03-19/reference_calendar_20260319T151315Z.csv
2021-04-05,2021,4,5,2,Abril,1,Segunda-feira,0,1,0,2026-03-19,2026-04-01T22:04:54.555Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/reference/reference_calendar/load_date=2026-03-19/reference_calendar_20260319T151315Z.csv



Null Value Analysis:


data,ano,mes,dia,trimestre,nome_mes,dia_semana,nome_dia_semana,is_fim_de_semana,is_inicio_mes,is_fim_mes,load_date,ingestion_timestamp,source_file
0,0,0,0,0,0,0,0,0,0,0,0,0,0



[INFO] Initial data overview completed successfully.


In [0]:
# ========================================
# 6. FACT GRAIN HYPOTHESIS
# ========================================


%md
### Fact Grain Definition — `fact_sales`

#### Overview
This section defines the granularity of the future Gold-layer fact table, ensuring alignment with dimensional modeling best practices.

#### Business Event
The central business event represented in the Gold layer is a **sales transaction at the order-item level**.

#### Confirmed Grain
**One row per order item (`item_pedido_id`).**

#### Primary Key
- `item_pedido_id` — Unique identifier of each order line.

#### Supporting Keys
- `pedido_id` — Order identifier
- `produto_id` — Product identifier
- `cliente_id` — Customer identifier
- `canal_id` — Sales channel identifier
- `data_pedido` — Order date

#### Rationale
- Represents the true transactional granularity of the ERP system.
- Ensures accurate aggregation of quantities, revenues, and costs.
- Prevents duplication and aggregation distortions.
- Aligns with dimensional modeling best practices.
- Supports Business Intelligence and Machine Learning workloads.

#### Source Datasets

| Layer | Dataset | Purpose |
|-------|---------|---------|
| Silver | `erp_order_items` | Transactional sales metrics |
| Silver | `silver_orders_customers` | Order and customer context |
| Silver | `erp_products` | Product attributes |
| Silver | `reference_sales_channels` | Sales channel dimension |
| Silver | `reference_calendar` | Time dimension |

In [0]:
# ========================================
# 7. ORDER ITEM PRIMARY KEY VALIDATION
# ========================================

from pyspark.sql import functions as F

print("=" * 80)
print("ORDER ITEM PRIMARY KEY VALIDATION")
print("=" * 80)

total_rows = df_order_items.count()
distinct_items = df_order_items.select("item_pedido_id").distinct().count()
null_items = df_order_items.filter(F.col("item_pedido_id").isNull()).count()

duplicate_items = (
    df_order_items
    .groupBy("item_pedido_id")
    .count()
    .filter(F.col("count") > 1)
    .count()
)

print(f"Total rows:                   {total_rows:,}")
print(f"Distinct item_pedido_id:      {distinct_items:,}")
print(f"Null item_pedido_id:          {null_items:,}")
print(f"Duplicate item_pedido_id:     {duplicate_items:,}")

if distinct_items == total_rows and null_items == 0:
    print("\n[SUCCESS] 'item_pedido_id' is a valid primary key.")
else:
    print("\n[WARNING] 'item_pedido_id' requires further validation.")

print("=" * 80)
print("[INFO] Order item primary key validation completed.")
print("=" * 80)

ORDER ITEM PRIMARY KEY VALIDATION
Total rows:                   245,287
Distinct item_pedido_id:      245,287
Null item_pedido_id:          0
Duplicate item_pedido_id:     0

[SUCCESS] 'item_pedido_id' is a valid primary key.
[INFO] Order item primary key validation completed.


In [0]:
# ========================================
# 8. HEADER–ITEM JOIN RECONSTRUCTION
# ========================================

from pyspark.sql import functions as F

print("=" * 80)
print("HEADER–ITEM JOIN RECONSTRUCTION")
print("=" * 80)

# Join between order items and enriched orders (customers)
df_header_item_join = df_order_items.alias("oi").join(
    df_orders_customers.alias("oc"),
    on="pedido_id",
    how="inner"
)

# Row count validation
order_items_count = df_order_items.count()
orders_customers_count = df_orders_customers.count()
joined_count = df_header_item_join.count()

print(f"Order Items Row Count:        {order_items_count:,}")
print(f"Orders Customers Row Count:   {orders_customers_count:,}")
print(f"Row Count After Join:         {joined_count:,}")

# Check for orphan order items
orphan_items = df_order_items.join(
    df_orders_customers,
    on="pedido_id",
    how="left_anti"
).count()

print(f"Orphan Order Items:           {orphan_items:,}")

if orphan_items == 0:
    print("[SUCCESS] No orphan order items detected.")
else:
    print("[WARNING] Orphan order items detected.")

print("=" * 80)
print("[INFO] Header-item join successfully reconstructed.")
print("=" * 80)

HEADER–ITEM JOIN RECONSTRUCTION
Order Items Row Count:        245,287
Orders Customers Row Count:   89,600
Row Count After Join:         245,287
Orphan Order Items:           0
[SUCCESS] No orphan order items detected.
[INFO] Header-item join successfully reconstructed.


### Null Value Analysis — Candidate Dataset for `fact_sales`

This analysis evaluates the presence of null values across all columns of the candidate dataset for the Gold-layer fact table.

#### Objectives
- Identify missing values in critical business keys and measures.
- Assess data quality before finalizing the fact table.
- Define appropriate data treatment strategies.
- Ensure referential integrity and analytical reliability.

#### Critical Columns for `fact_sales`

| Category | Columns |
|----------|---------|
| Primary Key | `item_pedido_id` |
| Foreign Keys | `pedido_id`, `produto_id`, `cliente_id`, `canal_id`, `data_pedido` |
| Measures | `quantidade`, `preco_venda_unitario`, `custo_unitario` |
| Supporting Attributes | `desconto_unitario`, `preco_lista_unitario` |

This step ensures that the Gold layer is built upon a reliable and well-governed dataset.

In [0]:
# ========================================
# 9. NULL VALUE ANALYSIS
# ========================================

print("=" * 80)
print("NULL VALUE ANALYSIS")
print("=" * 80)

df_fact_base = df_header_item_join
total_rows = df_fact_base.count()

print(f"Total Rows Analyzed: {total_rows:,}")

null_counts_df = df_fact_base.select([
    F.count(F.when(F.col(c).isNull(), c)).alias(c)
    for c in df_fact_base.columns
])

null_summary = (
    null_counts_df
    .selectExpr(
        f"stack({len(df_fact_base.columns)}, " +
        ", ".join([f"'{c}', `{c}`" for c in df_fact_base.columns]) +
        ") as (column_name, null_count)"
    )
    .withColumn("total_rows", F.lit(total_rows))
    .withColumn(
        "null_percentage",
        F.round((F.col("null_count") / F.col("total_rows")) * 100, 4)
    )
    .filter(F.col("null_count") > 0)
    .orderBy(F.col("null_count").desc())
)

display(null_summary)

print("=" * 80)
print("[INFO] Null value analysis completed successfully.")
print("=" * 80)

NULL VALUE ANALYSIS
Total Rows Analyzed: 245,287


column_name,null_count,total_rows,null_percentage
nota_item,216011,245287,88.0646
observacao_pedido,196303,245287,80.0299
obs_comercial,192484,245287,78.473
motivo_status,168413,245287,68.6596
vendedor_id,91352,245287,37.2429
produto_id,930,245287,0.3791


[INFO] Null value analysis completed successfully.


In [0]:
# ========================================
# NULL produto_id - IDENTIFICATION
# ========================================

from pyspark.sql import functions as F

df_null_products = df_fact_base.filter(F.col("produto_id").isNull())

null_count = df_null_products.count()
total_count = df_fact_base.count()

print("=" * 80)
print("NULL produto_id IDENTIFICATION")
print("=" * 80)
print(f"Null produto_id rows: {null_count:,}")
print(f"Total rows: {total_count:,}")
print(f"Percentage: {(null_count/total_count)*100:.4f}%")
print("=" * 80)

display(df_null_products.limit(10))

NULL produto_id IDENTIFICATION
Null produto_id rows: 930
Total rows: 245,287
Percentage: 0.3791%


pedido_id,item_pedido_id,produto_id,quantidade,preco_lista_unitario,desconto_unitario,preco_venda_unitario,custo_unitario,lote_fornecedor,flag_promocao,nota_item,load_date,ingestion_timestamp,source_file,cliente_id,data_pedido,canal_id,vendedor_id,cidade_entrega,estado_pedido,prazo_entrega_dias,observacao_pedido,sistema_origem,usuario_ultima_alteracao,order_load_date,order_ingestion_timestamp,order_source_file,nome_cliente,tipo_cliente,nif,data_registo,cliente_cidade,distrito,canal_captacao,score_atividade,obs_comercial,segmento,potencial_valor,cluster_comercial,status_cliente,data_status,motivo_status
PED2022041318507,IT000049144,null,5,8.55,0.0,8.55,5.27,L80571,0,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv,C0227,2022-04-13,CH02,VEND007,Braga,Entregue,4,null,ERP_FROZEN_V2,svc_sync,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Hotel 227,Hotel,253879650,2021-06-11,Braga,Braga,Vendas Externas,1.915,null,Grande,Médio,Cluster B,Ativo,2026-02-28,null
PED2022050419345,IT000051445,null,2,8.56,0.51,8.05,5.14,L27359,1,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv,C0072,2022-05-04,CH03,VEND008,Maia,Entregue,2,confirmar stock,ERP_FROZEN_V2,ana.c,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Supermercado 072,Supermercado,260744799,2021-07-25,Maia,Porto,Telefone,2.499,cliente estratégico,Grande,Alto,Cluster A,Ativo,2026-02-28,null
PED2022062721753,IT000058025,null,3,7.25,0.15,7.1,3.82,L30476,0,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv,C0232,2022-06-27,CH02,VEND007,Espinho,Entregue,4,null,ERP_FROZEN_V2,joao.p,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Supermercado 232,Supermercado,276705458,2022-04-21,Espinho,Aveiro,E-commerce,1.463,null,Médio,Baixo,Cluster B,Ativo,2026-02-28,null
PED2022110727701,IT000074197,null,3,8.11,0.73,7.38,4.07,L18361,1,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv,C0023,2022-11-07,CH03,null,Barcelos,Entregue,3,null,ERP_FROZEN_V2,joao.p,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Supermercado 023,Supermercado,269022083,2022-08-20,Barcelos,Braga,Vendas Externas,1.948,null,Médio,Baixo,Cluster A,Ativo,2026-02-28,null
PED2021072307635,IT000018823,null,5,4.56,0.32,4.24,2.44,L21462,1,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv,C0314,2021-07-23,CH02,null,Guimarães,Entregue,2,janela AM,ERP_FROZEN_V2,ana.c,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Restaurante 314,Restaurante,230293075,2021-07-06,Guimarães,Braga,E-commerce,2.24,null,Grande,Baixo,Cluster B,Ativo,2026-02-28,null
PED2021110211767,IT000030408,null,2,10.95,0.66,10.29,6.95,L13723,1,null,2026-03-17,2026-04-03T22:27:43.144Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_order_items/load_date=2026-03-17/erp_order_items.csv,C0260,2021-11-02,CH03,VEND008,Porto,Parcial,2,null,ERP_FROZEN_V2,maria.s,2026-03-17,2026-04-03T22:23:23.760Z,abfss://raw@stptfrozenfoodsdevwe01.dfs.core.windows.net/erp/erp_orders/load_date=2026-03-17/erp_orders.csv,Hotel 260,Hotel,215677431,2021-07-11,Porto,Porto,E-commerce,2.284,null,Grand

In [0]:
# ========================================
# NULL produto_id RECOVERY ANALYSIS
# ========================================

from pyspark.sql import functions as F
from pyspark.sql.window import Window

print("=" * 80)
print("NULL produto_id RECOVERY ANALYSIS")
print("=" * 80)

# --------------------------------------------------
# 1. PREPARE BASE DATA
# --------------------------------------------------

df_null_products = df_order_items.filter(F.col("produto_id").isNull())

df_valid_products = (
    df_products
    .select(
        "produto_id",
        "produto_nome",
        "categoria",
        "marca",
        "preco_lista_base",
        "custo_base_unitario",
        "fornecedor_id",
        "status_produto"
    )
    .dropDuplicates()
)

null_rows = df_null_products.count()
valid_products = df_valid_products.count()

print(f"Null produto_id rows to investigate:     {null_rows:,}")
print(f"Valid products available for matching:   {valid_products:,}")

# Optional enrichment from orders_customers
if "df_orders_customers" in locals():
    df_null_base = (
        df_null_products.alias("oi")
        .join(
            df_orders_customers.select(
                "pedido_id",
                "observacao_pedido",
                "cliente_id",
                "canal_id",
                "data_pedido"
            ).alias("oc"),
            on="pedido_id",
            how="left"
        )
    )
else:
    df_null_base = df_null_products

# --------------------------------------------------
# 2. NORMALIZE TEXT FIELDS
# --------------------------------------------------

def normalize_text(col_name):
    return F.trim(
        F.regexp_replace(
            F.lower(F.coalesce(F.col(col_name).cast("string"), F.lit(""))),
            r"[^a-zA-Z0-9À-ÿ ]",
            " "
        )
    )

df_null_base = (
    df_null_base
    .withColumn("nota_item_norm", normalize_text("nota_item"))
    .withColumn(
        "observacao_pedido_norm",
        normalize_text("observacao_pedido") if "observacao_pedido" in df_null_base.columns else F.lit("")
    )
    .withColumn(
        "texto_contexto",
        F.trim(
            F.concat_ws(
                " ",
                F.col("nota_item_norm"),
                F.col("observacao_pedido_norm")
            )
        )
    )
)

df_valid_products = (
    df_valid_products
    .withColumn("produto_nome_norm", normalize_text("produto_nome"))
    .withColumn("categoria_norm", normalize_text("categoria"))
    .withColumn("marca_norm", normalize_text("marca"))
)

# --------------------------------------------------
# 3. BUILD CANDIDATE MATCHES
# --------------------------------------------------
# Products table is very small (36 rows), so cross join is acceptable here.

df_candidates = (
    df_null_base.alias("n")
    .crossJoin(df_valid_products.alias("p"))
    .withColumn(
        "match_preco_venda_preco_lista",
        F.when(
            F.col("n.preco_venda_unitario") == F.col("p.preco_lista_base"),
            F.lit(1)
        ).otherwise(F.lit(0))
    )
    .withColumn(
        "match_custo",
        F.when(
            F.col("n.custo_unitario") == F.col("p.custo_base_unitario"),
            F.lit(1)
        ).otherwise(F.lit(0))
    )
    .withColumn(
        "match_preco_e_custo",
        F.when(
            (F.col("n.preco_venda_unitario") == F.col("p.preco_lista_base")) &
            (F.col("n.custo_unitario") == F.col("p.custo_base_unitario")),
            F.lit(1)
        ).otherwise(F.lit(0))
    )
    .withColumn(
        "match_nome_no_texto",
        F.when(
            (F.length(F.col("p.produto_nome_norm")) >= 4) &
            (F.instr(F.col("n.texto_contexto"), F.col("p.produto_nome_norm")) > 0),
            F.lit(1)
        ).otherwise(F.lit(0))
    )
    .withColumn(
        "match_marca_no_texto",
        F.when(
            (F.length(F.col("p.marca_norm")) >= 3) &
            (F.instr(F.col("n.texto_contexto"), F.col("p.marca_norm")) > 0),
            F.lit(1)
        ).otherwise(F.lit(0))
    )
    .withColumn(
        "match_categoria_no_texto",
        F.when(
            (F.length(F.col("p.categoria_norm")) >= 3) &
            (F.instr(F.col("n.texto_contexto"), F.col("p.categoria_norm")) > 0),
            F.lit(1)
        ).otherwise(F.lit(0))
    )
    .withColumn(
        "similarity_score",
        F.col("match_preco_e_custo") * F.lit(100) +
        F.col("match_preco_venda_preco_lista") * F.lit(40) +
        F.col("match_custo") * F.lit(30) +
        F.col("match_nome_no_texto") * F.lit(25) +
        F.col("match_marca_no_texto") * F.lit(10) +
        F.col("match_categoria_no_texto") * F.lit(5)
    )
    .withColumn(
        "match_rule",
        F.when(
            (F.col("match_preco_e_custo") == 1) & (F.col("match_nome_no_texto") == 1),
            F.lit("EXACT_PRICE_COST + NAME")
        ).when(
            (F.col("match_preco_e_custo") == 1) & (F.col("match_marca_no_texto") == 1),
            F.lit("EXACT_PRICE_COST + BRAND")
        ).when(
            F.col("match_preco_e_custo") == 1,
            F.lit("EXACT_PRICE_COST")
        ).when(
            (F.col("match_preco_venda_preco_lista") == 1) & (F.col("match_nome_no_texto") == 1),
            F.lit("PRICE + NAME")
        ).when(
            (F.col("match_custo") == 1) & (F.col("match_nome_no_texto") == 1),
            F.lit("COST + NAME")
        ).when(
            F.col("match_nome_no_texto") == 1,
            F.lit("NAME ONLY")
        ).when(
            F.col("match_marca_no_texto") == 1,
            F.lit("BRAND ONLY")
        ).otherwise(F.lit("NO RELEVANT MATCH"))
    )
)

df_relevant_candidates = df_candidates.filter(F.col("similarity_score") > 0)

# --------------------------------------------------
# 4. SELECT BEST CANDIDATES PER NULL ITEM
# --------------------------------------------------

df_relevant_candidates_flat = (
    df_relevant_candidates
    .select(
        F.col("n.item_pedido_id").alias("item_pedido_id"),
        F.col("n.pedido_id").alias("pedido_id"),
        F.col("n.quantidade").alias("quantidade"),
        F.col("n.preco_venda_unitario").alias("preco_venda_unitario"),
        F.col("n.custo_unitario").alias("custo_unitario"),
        F.col("n.nota_item").alias("nota_item"),
        F.col("n.observacao_pedido").alias("observacao_pedido"),
        F.col("n.texto_contexto").alias("texto_contexto"),
        F.col("p.produto_id").alias("produto_id_candidato"),
        F.col("p.produto_nome").alias("produto_nome_candidato"),
        F.col("p.marca").alias("marca_candidata"),
        F.col("p.categoria").alias("categoria_candidata"),
        F.col("p.preco_lista_base").alias("preco_lista_base_candidato"),
        F.col("p.custo_base_unitario").alias("custo_base_unitario_candidato"),
        F.col("p.fornecedor_id").alias("fornecedor_id_candidato"),
        F.col("similarity_score"),
        F.col("match_rule")
    )
)

w_best = Window.partitionBy("item_pedido_id").orderBy(
    F.col("similarity_score").desc(),
    F.col("produto_id_candidato").asc()
)

df_best_ranked = (
    df_relevant_candidates_flat
    .withColumn("rn", F.row_number().over(w_best))
    .withColumn(
        "best_score",
        F.max("similarity_score").over(Window.partitionBy("item_pedido_id"))
    )
)

df_best_candidates = df_best_ranked.filter(F.col("rn") == 1)

df_best_score_candidates = (
    df_relevant_candidates_flat.alias("c")
    .join(
        df_best_candidates.select("item_pedido_id", "best_score").alias("b"),
        (F.col("c.item_pedido_id") == F.col("b.item_pedido_id")) &
        (F.col("c.similarity_score") == F.col("b.best_score")),
        "inner"
    )
)

df_candidate_count = (
    df_best_score_candidates
    .groupBy("c.item_pedido_id")
    .agg(F.countDistinct("c.produto_id_candidato").alias("top_candidate_count"))
    .withColumnRenamed("item_pedido_id", "item_pedido_id_tmp")
)

df_final_recovery_view = (
    df_best_candidates
    .join(
        df_candidate_count,
        df_best_candidates["item_pedido_id"] == df_candidate_count["item_pedido_id_tmp"],
        "left"
    )
    .drop("item_pedido_id_tmp")
    .withColumn(
        "recovery_status",
        F.when(
            (F.col("similarity_score") >= 100) & (F.col("top_candidate_count") == 1),
            F.lit("HIGH CONFIDENCE")
        ).when(
            (F.col("similarity_score") >= 65) & (F.col("top_candidate_count") == 1),
            F.lit("MEDIUM CONFIDENCE")
        ).when(
            (F.col("similarity_score") > 0) & (F.col("top_candidate_count") > 1),
            F.lit("AMBIGUOUS")
        ).otherwise(F.lit("NO SAFE RECOVERY"))
    )
)

# --------------------------------------------------
# 5. BUILD WRITTEN SUMMARY
# --------------------------------------------------

high_confidence = df_final_recovery_view.filter(F.col("recovery_status") == "HIGH CONFIDENCE").count()
medium_confidence = df_final_recovery_view.filter(F.col("recovery_status") == "MEDIUM CONFIDENCE").count()
ambiguous = df_final_recovery_view.filter(F.col("recovery_status") == "AMBIGUOUS").count()

matched_any = df_final_recovery_view.count()
unmatched = null_rows - matched_any

price_cost_name = df_final_recovery_view.filter(F.col("match_rule") == "EXACT_PRICE_COST + NAME").count()
price_cost_brand = df_final_recovery_view.filter(F.col("match_rule") == "EXACT_PRICE_COST + BRAND").count()
price_cost_only = df_final_recovery_view.filter(F.col("match_rule") == "EXACT_PRICE_COST").count()
price_name = df_final_recovery_view.filter(F.col("match_rule") == "PRICE + NAME").count()
cost_name = df_final_recovery_view.filter(F.col("match_rule") == "COST + NAME").count()
name_only = df_final_recovery_view.filter(F.col("match_rule") == "NAME ONLY").count()
brand_only = df_final_recovery_view.filter(F.col("match_rule") == "BRAND ONLY").count()

print("=" * 80)
print("RECOVERY SUMMARY")
print("=" * 80)
print(f"Rows with null produto_id analyzed:              {null_rows:,}")
print(f"Rows with at least one possible candidate:       {matched_any:,}")
print(f"Rows with no relevant candidate found:           {unmatched:,}")
print("-" * 80)
print(f"High confidence recoveries:                      {high_confidence:,}")
print(f"Medium confidence recoveries:                    {medium_confidence:,}")
print(f"Ambiguous recoveries:                            {ambiguous:,}")
print("-" * 80)
print("Similarity patterns found:")
print(f"  EXACT_PRICE_COST + NAME:                       {price_cost_name:,}")
print(f"  EXACT_PRICE_COST + BRAND:                      {price_cost_brand:,}")
print(f"  EXACT_PRICE_COST:                              {price_cost_only:,}")
print(f"  PRICE + NAME:                                  {price_name:,}")
print(f"  COST + NAME:                                   {cost_name:,}")
print(f"  NAME ONLY:                                     {name_only:,}")
print(f"  BRAND ONLY:                                    {brand_only:,}")
print("=" * 80)

# --------------------------------------------------
# 6. SHOW TOP EXAMPLES
# --------------------------------------------------

print("[INFO] Sample of best candidate matches:")
display(
    df_final_recovery_view
    .orderBy(
        F.col("recovery_status").asc(),
        F.col("similarity_score").desc(),
        F.col("pedido_id").asc()
    )
    .limit(50)
)

print("[INFO] Recovery analysis completed successfully.")

NULL produto_id RECOVERY ANALYSIS
Null produto_id rows to investigate:     930
Valid products available for matching:   36
RECOVERY SUMMARY
Rows with null produto_id analyzed:              930
Rows with at least one possible candidate:       89
Rows with no relevant candidate found:           841
--------------------------------------------------------------------------------
High confidence recoveries:                      0
Medium confidence recoveries:                    0
Ambiguous recoveries:                            10
--------------------------------------------------------------------------------
Similarity patterns found:
  EXACT_PRICE_COST + NAME:                       0
  EXACT_PRICE_COST + BRAND:                      0
  EXACT_PRICE_COST:                              0
  PRICE + NAME:                                  0
  COST + NAME:                                   0
  NAME ONLY:                                     0
  BRAND ONLY:                                    0
[I

item_pedido_id,pedido_id,quantidade,preco_venda_unitario,custo_unitario,nota_item,observacao_pedido,texto_contexto,produto_id_candidato,produto_nome_candidato,marca_candidata,categoria_candidata,preco_lista_base_candidato,custo_base_unitario_candidato,fornecedor_id_candidato,similarity_score,match_rule,rn,best_score,top_candidate_count,recovery_status
IT000092443,PED2023040434382,8,5.9,3.4,null,null,,P012,Hambúrguer bovino 800g,Chef Express,Carne,5.9,3.32,F006,40,NO RELEVANT MATCH,1,40,4,AMBIGUOUS
IT000098790,PED2023052536697,3,3.4,2.11,troca embalagem,null,troca embalagem,P011,Mistura chinesa 1kg,Horta Fácil,Hortícolas,3.4,1.92,F004,40,NO RELEVANT MATCH,1,40,2,AMBIGUOUS
IT000221006,PED2025092480791,2,3.4,1.59,null,null,,P011,Mistura chinesa 1kg,Horta Fácil,Hortícolas,3.4,1.92,F004,40,NO RELEVANT MATCH,1,40,2,AMBIGUOUS
IT000015150,PED2021062206254,11,7.32,4.3,null,null,,P016,Empadão de carne 1200g,Atlântico Select,Pré-cozinhados,7.1,4.3,F001,30,NO RELEVANT MATCH,1,30,2,AMBIGUOUS
IT000026153,PED2021092310215,2,6.62,3.95,troca embalagem,null,troca embalagem,P003,Potas em anéis 1kg,Norte Mar,Peixe,6.9,3.95,F003,30,NO RELEVANT MATCH,1,30,2,AMBIGUOUS
IT000036515,PED2021121913953,7,6.23,4.3,null,null,,P016,Empadão de carne 1200g,Atlântico Select,Pré-cozinhados,7.1,4.3,F001,30,NO RELEVANT MATCH,1,30,2,AMBIGUOUS
IT000077117,PED2022112628762,1,7.32,4.3,null,null,,P016,Empadão de carne 1200g,Atlântico Select,Pré-cozinhados,7.1,4.3,F001,30,NO RELEVANT MATCH,1,30,2,AMBIGUOUS
IT000108248,PED2023080340113,4,6.92,4.25,null,null,,P006,Miolo de mexilhão 1kg,FrioMax,Marisco,7.4,4.25,F002,30,NO RELEVANT MATCH,1,30,2,AMBIGUOUS
IT000135934,PED2024022349962,1,7.0,4.3,troca embalagem,null,troca embalagem,P016,Empadão de carne 1200g,Atlântico Select,Pré-cozinhados,7.1,4.3,F001,30,NO RELEVANT MATCH,1,30,2,AMBIGUOUS
IT000151130,PED2024060955383,1,8.0,4.25,null,null,,P006,Miolo de mexilhão 1kg,FrioMax,Marisco,7.4,4.25,F002,30,NO RELEVANT MATCH,1,30,2,AMBIGUOUS


[INFO] Recovery analysis completed successfully.


%md
### Null `produto_id` Recovery Conclusion

A similarity-based recovery analysis was performed for the 930 records with null `produto_id`.

#### Result
- 930 records analyzed
- 89 records produced at least one possible candidate
- 0 high-confidence matches
- 0 medium-confidence matches
- 10 ambiguous matches
- 841 records with no relevant candidate

#### Conclusion
The available evidence is not sufficient to recover `produto_id` safely.

Potential matches were based mainly on price or cost similarity, without strong confirmation from product name, brand, or category context.

#### Decision
Records with null `produto_id` will not be imputed.

For the main `fact_sales` table, these records will be excluded to preserve referential integrity and analytical reliability.

If needed, they may be stored separately for audit and investigation purposes.

In [0]:
# ========================================
# 10. NULL produto_id FINANCIAL IMPACT
# ========================================

print("=" * 80)
print("NULL produto_id FINANCIAL IMPACT")
print("=" * 80)
print("[INFO] Calculating the business impact of rows with null produto_id...")
print("[INFO] Comparing excluded rows against the full order items dataset.")

# Impact of null produto_id rows
impact_null = df_null_products.agg(
    F.sum("quantidade").alias("null_total_quantity"),
    F.sum(F.col("quantidade") * F.col("preco_venda_unitario")).alias("null_total_revenue"),
    F.sum(F.col("quantidade") * F.col("custo_unitario")).alias("null_total_cost")
).collect()[0]

# Impact of full dataset
impact_full = df_order_items.agg(
    F.sum("quantidade").alias("full_total_quantity"),
    F.sum(F.col("quantidade") * F.col("preco_venda_unitario")).alias("full_total_revenue"),
    F.sum(F.col("quantidade") * F.col("custo_unitario")).alias("full_total_cost")
).collect()[0]

null_total_quantity = impact_null["null_total_quantity"] or 0
null_total_revenue = impact_null["null_total_revenue"] or 0
null_total_cost = impact_null["null_total_cost"] or 0

full_total_quantity = impact_full["full_total_quantity"] or 0
full_total_revenue = impact_full["full_total_revenue"] or 0
full_total_cost = impact_full["full_total_cost"] or 0

quantity_pct = round((null_total_quantity / full_total_quantity) * 100, 4) if full_total_quantity else 0
revenue_pct = round((null_total_revenue / full_total_revenue) * 100, 4) if full_total_revenue else 0
cost_pct = round((null_total_cost / full_total_cost) * 100, 4) if full_total_cost else 0

print("[INFO] Financial impact summary:")
print(f"Null produto_id rows:                     {df_null_products.count():,}")
print(f"Total order item rows:                   {df_order_items.count():,}")
print("-" * 80)
print(f"Quantity in null produto_id rows:        {null_total_quantity:,.2f}")
print(f"Total quantity in full dataset:          {full_total_quantity:,.2f}")
print(f"Quantity impact (%):                     {quantity_pct}%")
print("-" * 80)
print(f"Revenue in null produto_id rows:         {null_total_revenue:,.2f}")
print(f"Total revenue in full dataset:           {full_total_revenue:,.2f}")
print(f"Revenue impact (%):                      {revenue_pct}%")
print("-" * 80)
print(f"Cost in null produto_id rows:            {null_total_cost:,.2f}")
print(f"Total cost in full dataset:              {full_total_cost:,.2f}")
print(f"Cost impact (%):                         {cost_pct}%")
print("=" * 80)
print("[INFO] Financial impact analysis completed successfully.")
print("=" * 80)

NULL produto_id FINANCIAL IMPACT
[INFO] Calculating the business impact of rows with null produto_id...
[INFO] Comparing excluded rows against the full order items dataset.
[INFO] Financial impact summary:
Null produto_id rows:                     930
Total order item rows:                   245,287
--------------------------------------------------------------------------------
Quantity in null produto_id rows:        3,382.00
Total quantity in full dataset:          918,175.00
Quantity impact (%):                     0.3683%
--------------------------------------------------------------------------------
Revenue in null produto_id rows:         20,619.21
Total revenue in full dataset:           5,507,958.47
Revenue impact (%):                      0.3744%
--------------------------------------------------------------------------------
Cost in null produto_id rows:            12,497.41
Total cost in full dataset:              3,340,593.90
Cost impact (%):                         0.374

In [0]:
# ========================================
# NULL produto_id ENRICHED BASE
# ========================================

print("=" * 80)
print("NULL produto_id ENRICHED BASE")
print("=" * 80)
print("[INFO] Building enriched dataset for null produto_id analysis...")
print("[INFO] Joining order items with orders and customers to obtain context.")

# Join between order items and orders/customers
df_header_item_join = df_order_items.alias("oi").join(
    df_orders_customers.alias("oc"),
    on="pedido_id",
    how="inner"
)

# Filter rows with null produto_id
df_null_products_enriched = df_header_item_join.filter(
    F.col("oi.produto_id").isNull()
)

NULL produto_id ENRICHED BASE
[INFO] Building enriched dataset for null produto_id analysis...
[INFO] Joining order items with orders and customers to obtain context.


In [0]:
# ========================================
# 11. NULL produto_id DISTRIBUTION BY SALES CHANNEL
# ========================================

print("=" * 80)
print("NULL produto_id DISTRIBUTION BY SALES CHANNEL")
print("=" * 80)
print("[INFO] Analyzing whether null produto_id rows are concentrated in specific sales channels...")
print("[INFO] This helps identify whether the issue is localized or broadly distributed.")

df_null_by_channel = (
    df_null_products_enriched
    .groupBy("canal_id")
    .count()
    .withColumnRenamed("count", "null_produto_id_rows")
    .orderBy(F.desc("null_produto_id_rows"))
)

print(f"[INFO] Distinct sales channels affected: {df_null_by_channel.count():,}")
display(df_null_by_channel)

print("=" * 80)
print("[INFO] Sales channel distribution analysis completed successfully.")
print("=" * 80)

NULL produto_id DISTRIBUTION BY SALES CHANNEL
[INFO] Analyzing whether null produto_id rows are concentrated in specific sales channels...
[INFO] This helps identify whether the issue is localized or broadly distributed.
[INFO] Distinct sales channels affected: 4


canal_id,null_produto_id_rows
CH02,387
CH03,227
CH01,219
CH04,97


[INFO] Sales channel distribution analysis completed successfully.


In [0]:
# ========================================
# 12. NULL produto_id DISTRIBUTION BY CUSTOMER
# ========================================

print("=" * 80)
print("NULL produto_id DISTRIBUTION BY CUSTOMER")
print("=" * 80)
print("[INFO] Analyzing whether null produto_id rows are concentrated in specific customers...")
print("[INFO] This helps determine whether the issue is isolated or recurrent across the customer base.")

df_null_by_customer = (
    df_null_products_enriched
    .groupBy("cliente_id")
    .count()
    .withColumnRenamed("count", "null_produto_id_rows")
    .orderBy(F.desc("null_produto_id_rows"))
)

distinct_customers_with_nulls = df_null_products_enriched.select("cliente_id").distinct().count()

print(f"[INFO] Distinct customers affected: {distinct_customers_with_nulls:,}")
display(df_null_by_customer)

print("=" * 80)
print("[INFO] Customer distribution analysis completed successfully.")
print("=" * 80)

NULL produto_id DISTRIBUTION BY CUSTOMER
[INFO] Analyzing whether null produto_id rows are concentrated in specific customers...
[INFO] This helps determine whether the issue is isolated or recurrent across the customer base.
[INFO] Distinct customers affected: 101


cliente_id,null_produto_id_rows
C0164,65
C0227,40
C0072,40
C0260,37
C0043,34
C0314,33
C0057,32
C0004,30
C0136,29
C0103,29


[INFO] Customer distribution analysis completed successfully.


In [0]:
# ========================================
# 13. NULL produto_id TEMPORAL DISTRIBUTION
# ========================================

print("=" * 80)
print("NULL produto_id TEMPORAL DISTRIBUTION")
print("=" * 80)
print("[INFO] Analyzing the yearly distribution of null produto_id rows...")
print("[INFO] This helps identify whether the issue is persistent across time or concentrated in specific periods.")

df_null_by_year = (
    df_null_products_enriched
    .groupBy(F.year("data_pedido").alias("ano"))
    .count()
    .withColumnRenamed("count", "null_produto_id_rows")
    .orderBy("ano")
)

display(df_null_by_year)

print("=" * 80)
print("[INFO] Temporal distribution analysis completed successfully.")
print("=" * 80)

NULL produto_id TEMPORAL DISTRIBUTION
[INFO] Analyzing the yearly distribution of null produto_id rows...
[INFO] This helps identify whether the issue is persistent across time or concentrated in specific periods.


ano,null_produto_id_rows
2021,152
2022,142
2023,190
2024,185
2025,229
2026,32


[INFO] Temporal distribution analysis completed successfully.


In [0]:
# ========================================
# 14. NULL produto_id TEXT FIELD INSPECTION
# ========================================

print("=" * 80)
print("NULL produto_id TEXT FIELD INSPECTION")
print("=" * 80)
print("[INFO] Inspecting descriptive text fields for possible recovery clues...")
print("[INFO] The goal is to identify whether free-text columns contain evidence of the missing product.")
print("[INFO] Typical clues may include product names, packaging changes, urgency notes, or commercial comments.")

df_null_text_inspection = df_null_products_enriched.select(
    "pedido_id",
    "item_pedido_id",
    "nota_item",
    "observacao_pedido",
    "obs_comercial"
)

print("[INFO] Displaying sample records with descriptive text fields:")
df_null_text_inspection.show(20, truncate=False)

print("=" * 80)
print("[INFO] Text field inspection completed successfully.")
print("=" * 80)

NULL produto_id TEXT FIELD INSPECTION
[INFO] Inspecting descriptive text fields for possible recovery clues...
[INFO] The goal is to identify whether free-text columns contain evidence of the missing product.
[INFO] Typical clues may include product names, packaging changes, urgency notes, or commercial comments.
[INFO] Displaying sample records with descriptive text fields:
+----------------+--------------+--------------+-----------------+-------------------+
|pedido_id       |item_pedido_id|nota_item     |observacao_pedido|obs_comercial      |
+----------------+--------------+--------------+-----------------+-------------------+
|PED2022041318507|IT000049144   |NULL          |NULL             |NULL               |
|PED2022050419345|IT000051445   |NULL          |confirmar stock  |cliente estratégico|
|PED2022062721753|IT000058025   |NULL          |NULL             |NULL               |
|PED2022110727701|IT000074197   |NULL          |NULL             |NULL               |
|PED202107230

%md
### Null `produto_id` — Final Decision

A comprehensive analysis was conducted on the 930 records with null `produto_id`.

#### Key Findings

- **Financial Impact**
  - Quantity Impact: **0.3683%**
  - Revenue Impact: **0.3744%**
  - Cost Impact: **0.3741%**

- **Sales Channel Distribution**
  - Present across all channels, with no concentration in a specific source.

- **Customer Distribution**
  - Affects 101 distinct customers, indicating a dispersed issue.

- **Temporal Distribution**
  - Occurs consistently between 2021 and 2026, suggesting historical inconsistencies in source data.

- **Recovery Analysis**
  - No high-confidence matches were identified using similarity-based techniques.

#### Decision

Given the minimal financial impact and the inability to recover the missing product identifiers with confidence, the records with null `produto_id` will be excluded from the analytical fact table.

#### Justification

This decision ensures:

- Referential integrity with the product dimension;
- Reliability of business metrics and KPIs;
- Alignment with enterprise data quality standards;
- Consistency with dimensional modeling best practices.

If required, these records may be stored in a separate audit dataset for traceability.

%md
### Gold Layer Exploratory Conclusion

This exploratory notebook validated the datasets required to build the Gold layer of the PT Frozen Foods data platform. The objective was to confirm data quality, validate dimensional assumptions, and ensure readiness for analytical consumption.

---

### Grain Validation

The transactional grain for the `fact_sales` table was confirmed as:

- **One record per order item (`item_pedido_id`).**

Validation results:

- Total rows: 245,287
- Distinct `item_pedido_id`: 245,287
- Null values: 0
- Duplicate values: 0

This confirms that `item_pedido_id` is a valid primary key and ensures the integrity of the fact table.

---

### Null `produto_id` Assessment

A total of 930 records were identified with null `produto_id`. A comprehensive analysis was conducted to evaluate their impact and potential recovery.

#### Financial Impact

- Quantity Impact: 0.3683%
- Revenue Impact: 0.3744%
- Cost Impact: 0.3741%

#### Distribution Analysis

- **Sales Channels:** Distributed across all channels.
- **Customers:** Affecting 101 distinct customers.
- **Time Period:** Spread between 2021 and 2026.

#### Recovery Analysis

Similarity-based matching using price, cost, and contextual attributes did not produce reliable results. No high-confidence recovery was possible.

#### Decision

To ensure referential integrity and analytical reliability, records with null `produto_id` will be excluded from the Gold layer.

---

### Dimensional Model Alignment

The exploratory analysis confirmed alignment with the dimensional model defined in the project documentation.

Validated entities:

- `fact_sales`
- `dim_product`
- `dim_customer`
- `dim_supplier`
- `dim_sales_channel`
- `dim_calendar`

Detailed specifications are documented in **12_data_model.md**.

---

### Medallion Architecture Alignment

The datasets are consistent with the Medallion Architecture:

| Layer | Status |
|-------|--------|
| Bronze | Completed |
| Silver | Completed |
| Gold | Validated |

---

### Governance and Technology Alignment

| Component | Technology |
|-----------|------------|
| Storage | ADLS Gen2 |
| Format | Delta Lake |
| Processing | Azure Databricks |
| Governance | Unity Catalog |
| Orchestration | Azure Data Factory |

These standards are defined in the project architecture documentation.

---

### Key Decisions

| Decision | Outcome |
|----------|---------|
| Fact table grain | Defined as `item_pedido_id` |
| Null `produto_id` handling | Excluded due to minimal impact |
| Dimensional modeling | Star Schema adopted |
| Referential integrity | Validated and enforced |
| Gold layer readiness | Confirmed |

---

### Conclusion

The datasets have been validated and approved for Gold layer processing. The analysis ensures:

- data quality and integrity;
- adherence to dimensional modeling best practices;
- scalability for business intelligence and machine learning;
- alignment with enterprise Lakehouse standards.

The platform is now ready for the implementation of the Gold processing notebooks.